core/src/parking_lot.rs

```diff
@@ -263,14 +277,19 @@ fn create_hashtable() -> &'static HashTable {
 // created, which only happens once per thread.
 fn grow_hashtable(num_threads: usize) {
     // Lock all buckets in the existing table and get a reference to it
-    let old_table = loop {
+    let (old_table, mut new_table) = loop {
         let table = get_hashtable();
 
         // Check if we need to resize the existing table
         if table.entries.len() >= LOAD_FACTOR * num_threads {
             return;
         }
 
+        // For feature = "reserve", create the new table here, before we lock
+        // all the buckets, in case the allocator needs to acquire a lock.
+        #[cfg(feature = "reserve")]
+        let new_table = HashTable::new(num_threads, table);
```

Comment:
This doesn't sound right: if the allocator attempts to take a lock at this point, won't it just trigger hashtable growth again? This would result in infinite recursion instead of a deadlock, which isn't much better.

Reply:
If the user calls `reserve()` before all new threads, then `grow_hashtable` will only ever run on threads which already have enough hashtable space for themselves, so the allocator can take locks here.

Reply:
Except that there's an issue with calling `reserve` before creating a thread: thread creation may fail, but there's no way to undo the reservation. Every time creating a thread fails, `NUM_RESERVED` will still be increased.

If you instead call `reserve` as the first action within a new thread then you run into the problem I described above where the allocator can cause infinite recursion.

I think the correct approach here is to do hashtable growth as the first thing in a thread, while using a thread-local `bool` to detect recursion and skip the growth if it happens within the allocator. This is fine because the hash table here uses chaining so its capacity is not limited to the number of buckets.